<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <title>consexpression - Costa-Silva et al. (2017)</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="Clean responsive bootstrap website template">
    <meta name="author" content="">
    <!-- styles -->
    <link href="assets/css/bootstrap.css" rel="stylesheet">
    <link href="assets/css/bootstrap-responsive.css" rel="stylesheet">
    <link href="assets/css/docs.css" rel="stylesheet">
    <link href="assets/css/prettyPhoto.css" rel="stylesheet">
    <link href="assets/js/google-code-prettify/prettify.css" rel="stylesheet">
    <link href="assets/css/flexslider.css" rel="stylesheet">
    <link href="assets/css/refineslide.css" rel="stylesheet">
    <link href="assets/css/font-awesome.css" rel="stylesheet">
    <link href="assets/css/animate.css" rel="stylesheet">
    <link href="https://fonts.googleapis.com/css?family=Open+Sans:400italic,400,600,700" rel="stylesheet">

    <link href="assets/css/style.css" rel="stylesheet">
    <link href="assets/color/default.css" rel="stylesheet">

    <!-- fav and touch icons -->
    <link rel="shortcut icon" href="assets/ico/favicon.ico">
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="assets/ico/apple-touch-icon-144-precomposed.png">
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="assets/ico/apple-touch-icon-114-precomposed.png">
    <link rel="apple-touch-icon-precomposed" sizes="72x72" href="assets/ico/apple-touch-icon-72-precomposed.png">
    <link rel="apple-touch-icon-precomposed" href="assets/ico/apple-touch-icon-57-precomposed.png">

    <!-- =======================================================
      Theme Name: Plato
      Theme URL: https://bootstrapmade.com/plato-responsive-bootstrap-website-template/
      Author: BootstrapMade.com
      Author URL: https://bootstrapmade.com
    ======================================================= -->
</head>

<body>
<header>
    <!-- Navbar
    ================================================== -->
    <div class="cbp-af-header">
        <div class=" cbp-af-inner">
            <div class="container">
                <div class="row">

                    <div class="span4">
                        <!-- logo -->
                        <div class="logo">
                            <h1><a href="index.html">consexpression</a></h1>
                            <!-- <img src="assets/img/logo.png" alt="" /> -->
                        </div>
                        <!-- end logo -->
                    </div>
                    <!-- top menu -->
                    <div class="span8">

                        <div class="navbar">
                            <div class="navbar-inner">
                                <nav>
                                    <ul class="nav topnav">
                                        <li class="dropdown active">
                                            <a href="index.html">Home</a>
                                        </li>
                                        <li class="dropdown">
                                            <a href="#">Features</a>
                                            <!--
                                            <ul class="dropdown-menu">
                                                <li><a href="#">Mapping</a></li>
                                                <li><a href="#">Count read alignment</a></li>
                                                <li><a href="#">Expression</a></li>
                                                <li><a href="#">Tutorial</a></li>

                                            </ul>
                                            -->
                                        </li>
                                        <li class="dropdown">
                                            <a href="#install">Install</a>
                                        </li>
                                        <li class="dropdown">
                                            <a href="#mapdge">Tutorial</a>
                                            <!--
                                            <ul class="dropdown-menu">
                                                <li><a href="#">Mapping and Expression</a></li>
                                                <li><a href="#">Only Expression Analysis</a></li>
                                            </ul>
                                            -->
                                        </li>
                                    </ul>
                                </nav>
                            </div>
                        </div>
                        <!-- end menu -->
                    </div>

                </div>
            </div>
        </div>
    </div>
</header>
<section id="intro">

    <div class="container">
        <div class="row">
            <div class="span12">
                <h2><strong>Consensus among <span class="highlight primary">five DEGs</span> identification
                    methods</strong></h2>
                <p class="lead">
                    Diferential Expression Analysis, with a consensus result.
                </p>
                <ul class="list list-ok strong bigger">
                    <li>Mapping, count and DEGs Analysys in one tool</li>
                    <li>Valid HTML5 code and well structured</li>
                </ul>
            </div>
        </div>
    </div>
</section>

<section id="maincontent">
    <div class="container">
        <div class="row">
            <div class="span4">
                <div class="features">
                    <div class="icon">
                        <i class="icon-bg-light icon-circled icon-download-alt icon-5x active"></i>
                    </div>
                    <div class="features_content">
                        <h3>Install</h3>
                        <p class="left">
                            Setting up <strong>consexpression</strong> takes just a few commands.
                            In <a href="#install"><strong>install</strong></a> section you can see ste by step to run
                            consexpression.
                        </p>
                        <!-- <a href="#" class="btn btn-color btn-rounded"><i class="icon-angle-right"></i> Read more</a> -->
                    </div>
                </div>
            </div>
            <div class="span4">
                <div class="features">
                    <div class="icon">
                        <i class="icon-bg-dark icon-circled icon-terminal icon-5x"></i>
                    </div>
                    <div class="features_content">
                        <h3>Cross platform</h3>
                        <p class="left">
                            We use Python and R language in development of consexpression. You can run Consexpression on
                            Windows, Mac OS and Linux.
                        </p>
                        <!-- <a href="#" class="btn btn-color btn-rounded"><i class="icon-angle-right"></i> Read more</a> -->
                    </div>
                </div>
            </div>
            <div class="span4">
                <div class="features">
                    <div class="icon">
                        <i class="icon-bg-dark icon-circled icon-quote-left icon-5x"></i>
                    </div>
                    <div class="features_content">
                        <h3>Read the full article</h3>
                        <p class="left">
                            The complete study was be published in decemebr, 2017. PLOS ONE is an journal community
                            working together to advance science for the benefit of society. </a>
                        </p>
                        <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0190152"
                           class="btn btn-color btn-rounded"><i class="icon-angle-right"></i> Read the full article</a>
                    </div>
                </div>
            </div>

        </div>

        <!-- blank divider -->
        <div class="row">
            <div class="span12">
                <div class="blank10"></div>
            </div>
        </div>
</section>
<section id="install">
    <div class="container">
        <div class="row">

            <div class="span12">

                <div class="page-header">
                    <h3>Install consexpression</h3>
                </div>
                <h5>Requirements</h5>
                <p>
                    Consexpression makes use of various tools and, have various requirements:

                </p>
                <ul>
                    <li><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">Bowtie 2</a> is a tool for
                        aligning sequencing reads.
                    </li>
                    <li><a href="https://ccb.jhu.edu/software/tophat/index.shtml">TopHat 2</a>, is a splice junction
                        mapper for RNA-Seq reads.
                    </li>
                    <li><a href="https://www.r-project.org/">R language</a>, is a free software environment for
                        statistical computing and graphics.
                    </li>
                    <li><a href="https://www.python.org/">Python</a>, is a programming language that lets you work more
                        quickly and integrate your systems more effectively.
                    </li>
                    <li><a href="https://rpy2.github.io/doc/latest/html/index.html">rpy2</a>, is a interface between R
                        an Python languages to benefit from the libraries of one language while working in the other.
                    </li>
                    <li><a href="https://htseq.readthedocs.io/en/release_0.11.1/">HTSeq</a>, is a Python framework to
                        work with high-throughput sequencing data.
                    </li>
                </ul>

                <!-- blank divider -->
                <div class="row">
                    <div class="span12">
                        <div class="blank10"></div>
                    </div>
                </div>


                <div class="tab-content">
                    <div class="tab-pane active" id="general">
                        <p>
                            If you have correctly install all requirements, follow the steps below to install <strong>consexpression</strong>:
                        </p>

                        <p>
                            <i class="icon-terminal icon-1x"></i> Access the <strong>R</strong> workspace via the
                            terminal: by typing:
                        <div class="control-group success"><label class="control-label">R</label></div>
                        You can also access the R workspace through an IDE like RStudio.
                        In R workspace, follow commands:
                        </p>
                        <pre class="prettyprint linenums">
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("DESeq")
BiocManager::install("edgeR")
BiocManager::install("baySeq")
BiocManager::install("EBSeq")
BiocManager::install("samr")
BiocManager::install("NOISeq")
BiocManager::install("limma")
BiocManager::install("DESeq2")
BiocManager::install("devtools")
BiocManager::install("pachterlab/sleuth")</pre>
                        <small>* These commands was tested in R language version 3.6.2</small>
                    </div>
                    <!-- blank divider -->
                    <div class="row">
                        <div class="span12">
                            <div class="blank10"></div>
                        </div>
                    </div>
                    <p>Download or clone consexpression</p>

                    <i class="icon-terminal icon-1x"></i> In terminal:
                    <p>Go to folder where you like put consexpression code:</p>
                    <pre class="prettyprint linenums">
cd my_folder
git clone git@github.com:costasilvati/consexpression.git
                            </pre>


                    <!-- blank divider -->
                    <div class="row">
                        <div class="span12">
                            <div class="blank10"></div>
                        </div>
                    </div>
                    <h5 id="headings">Specific Instructions</h5>
                    <div class="accordion" id="accordion2">
                        <div class="accordion-group">
                            <div class="accordion-heading">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2"
                                   href="#linux">
                                    <i class="icon-caret-right"></i> Linux instructions <i
                                        class="icon-linux icon-2x"></i> </a>
                            </div>
                            <div id="linux" class="accordion-body collapse">
                                <div class="accordion-inner">
                                    Working...
                                </div>
                            </div>
                        </div>
                        <div class="accordion-group">
                            <div class="accordion-heading">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2"
                                   href="#windows">
                                    <i class="icon-caret-right"></i> Windows instructions <i
                                        class="icon-windows icon-2x"></i> </a>
                            </div>
                            <div id="windows" class="accordion-body collapse">
                                <div class="accordion-inner">
                                    Working...
                                </div>
                            </div>
                        </div>
                        <div class="accordion-group">
                            <div class="accordion-heading">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2"
                                   href="#mac">
                                    <i class="icon-caret-right"></i> Mac OS instructions <i
                                        class="icon-apple icon-2x"></i> </a>
                            </div>
                            <div id="mac" class="accordion-body collapse">
                                <div class="accordion-inner">
                                    Working...
                                </div>
                            </div>
                        </div>
                    </div>

                </div>
                <section id="mapdge">
                    <!-- blank divider -->
                    <div class="row">
                        <div class="span12">
                            <div class="blank10"></div>
                        </div>
                    </div>
                    <div class="container">
                        <div class="row">

                            <div class="span12">

                                <div class="page-header">
                                    <h2>Using consexpression</h2>
                                </div>
                                <h5>Configure your analyse </h5>
                                <p>
                                    Consexpression use a file to configure analyse, this file is located in
                                    <span>dao</span> folder, named CONFIG_tool.txt.
                                    This file is a parameter used to execute consexpression.
                                </p>
                                <div class="bs-docs-separator">
                                    <table class="table">
                                        <thead>
                                        <tr>
                                            <th>Paremeter</th>
                                            <th>Information expected</th>
                                            <th>Usage</th>
                                        </tr>
                                        </thead>
                                        <tbody>
                                        <tr>
                                            <td> #===</td>
                                            <td>Section of file</td>
                                            <td>All line start with '#' do not is processed</td>
                                        </tr>
                                        <tr>
                                            <td>NAME:</td>
                                            <td>Name to identify results</td>
                                            <td>Will be used like prefix or suffix in generated files</td>
                                        </tr>
                                        <tr>
                                            <td>REPLIC:</td>
                                            <td>Integer number of biological replicates in each treatment to analyse</td>
                                            <td>This nunmer define how columns are expected in table count</td>
                                        </tr>
                                        <tr>
                                            <td>GROUP_NUMBER:</td>
                                            <td>Whole number indicates number of treatments</td>
                                            <td>Is used for compare expression betwen groups.</td>
                                        </tr>
                                        <tr>
                                            <td>GROUP_NAMES:</td>
                                            <td>Comma-separated list with the name of the treatments</td>
                                            <td>Used for make identify replics and treatment</td>
                                        </tr>
                                        <tr>
                                            <td>REFERENCE_GENOME:</td>
                                            <td>Path to FASTA file with complete genome of organism to mapping reads.</td>
                                            <td>Used for align reads</td>
                                        </tr>
                                        <tr>
                                            <td>READS_DIRECTORY:</td>
                                            <td>Absolute path of folder treatments.</td>
                                            <td>Used to identify the location of all FASTQs in the analysis.
                                                Within this folder there must be sub-folders separated by treatments (groups).</td>
                                        </tr>
                                        <tr>
                                            <td>GROUP_DIRECTORIES:</td>
                                            <td>Comma-separated list, with the absolute path of folders with FASTQ files. One folder per group.</td>
                                            <td>It is expected to find a folder for each treatment.
                                                Each folder must contain the same number of FASTQ files, indicated in replicates.</td>
                                        </tr>
                                        <tr>
                                            <td>PAIRED_END:</td>
                                            <td>Boolean information [True|TRUE|False|FALSE]</td>
                                            <td>Used to identify sequencing RNA-Seq method - <span>consexpression don't make paired end analyse</span></td>
                                        </tr>
                                        <tr>
                                            <td>THREADS:</td>
                                            <td>Integer</td>
                                            <td>Used to parameterize processing usage by the mapping tool.</td>
                                        </tr>
                                        <tr>
                                            <td>MODE:</td>
                                            <td>Text, can be one of the options: "union" or "intersection-strict" or "intersection-nonempty".</td>
                                            <td>Used by the htseq-count tool, it defines the way to consider the count of mapped reads.
                                                Read more at:
                                                <a href="https://htseq.readthedocs.io/en/release_0.11.1/counting.html">documentation page.</a>
                                            </td>
                                        </tr>
                                        <tr>
                                            <td>ANOTATION_FILE:</td>
                                            <td>Absolute path of the GTF file linked to the genome file informed in REFERENCE.</td>
                                            <td>Used by the htseq-count tool to generate conversion tables based on SAM mapping files.</td>
                                        </tr>
                                        <tr>
                                            <td>OUTPUT:</td>
                                            <td>Absolute path to save results of expression analysis.</td>
                                            <td>Used to write output files for differential expression analysis tools. The file with the genes identified by consensus is also written in this folder.</td>
                                        </tr>
                                        </tbody>
                                    </table>
                                </div>
                                <p>With the configuration file filled in, navigate via terminal to the consexpression installation folder, and type:</p>
                                <pre class="prettyprint linenums">
python experiment.py dao/CONFIG_tool.txt
                                </pre>
                                <div class="alert">
                                <button type="button" class="close" data-dismiss="alert">&times;</button>
                                <strong>Warning!</strong>
                                    The configuration file is available with the tool in the project's dao folder, named CONFGI_tool.txt.
                                    This file can be in other locations and be renamed, as long as it follows the original training standard.
                                    If it is necessary to use this file in another location, remember to inform the file's absolute path when
                                    running the tool, replacing <span> dao / CONFIG_toll.txt </span> with the path of your file.

                                </div>

                            </div>
                        </div>
                        <div class="row">
                            <div class="span12">

                                <div class="page-header">
                                    <h5>Perform expression analysis (consensus) only</h5>
                                </div>

                                <p>This tool was developed to perform a complete analysis: Mapping, counting, expression analysis and consensus.</p>
                                <p>The need was identified for many users who had count data and wished to identify only the consensus of their results.</p>
                                <p>To perform expression analysis only, you should normally fill in the configuration file with all the information from your experiment.
                                    Leave only the REFERENCE field of the configuration file empty. In this way, consexpression will identify that there is no need to perform
                                    mapping and contact, and will do the expression analysis.</p>
                                <p>Run the tool normally, with the command:</p>
                                <pre class="prettyprint linenums">
python experiment.py dao/CONFIG_tool.txt
                                </pre>
                                <p>During execution the tool will ask for the absolute path of the counting table (text file, separated by a comma),
                                    type the path of the file and press enter.
                                    The analysis of expression and consensus will be performed.</p>
                            </div>
                        </div>
                    </div>
                </section>

                <!--<pre class="prettyprint linenums">
                         &lt;!DOCTYPE html&gt; &lt;html lang="en"&gt; ... &lt;/html&gt;
                </pre>
                <h5>Typography and links</h5>
                <p>
                    Since Serenity template built with bootstrap so we are using bootstrap basic global display, typography, and link styles. In addition we add our custom rules to remove radius, shadow and other things by adding 'overwrite.css' file in folder CSS
                </p>
                <p>
                    All basic bootstrap components style also changed with 'overwrite.css' file such as buttons, navigation, menu..etc
                </p>-->

            </div> <!-- span12 -->
        </div><!-- row -->
    </div> <!-- class="container -->
</section>
<!-- Footer
================================================== -->
<footer class="footer">
    <div class="container">
        <div class="row">
            <div class="span4">
                <div class="widget">
                    <!-- logo -->
                    <div class="footerlogo">
                        <h6><a href="index.html">consexpression</a></h6>
                        <!-- <img src="assets/img/logo.png" alt="" /> -->
                    </div>

                    <!-- end logo -->
                    <address>
                        <strong>Costa-Silva, et. al, 2017.</strong><br>
                        Department of Computer Science, Bioinformatics Graduate Program, Federal University of
                        Technology<br>
                        Paraná, Cornélio Procópio, PR, Brazil,<br>
                    </address>
                    <i class="icon-envelope"></i> julianacostasilvati@gmail.com <br>
                    <i class="icon-envelope"></i> fabricio.m.lopes@utpfr.edu.br

                </div>
            </div>
            <div class="span3">
                <div class="widget">
                    <h5>Browse pages</h5>

                </div>
            </div>
            <div class="span3">
                <div class="widget">
                    <h5>Keep updated</h5>
                    <p>
                        Watch in github <strong> <a href="https://github.com/costasilvati/consexpression">
                        repository</a> </strong>
                    </p>

                    <ul class="social-network">
                        <li><a href="#"><i class="icon-bg-light icon-facebook icon-circled icon-1x"></i></a></li>
                        <li><a href="#" title="Twitter"><i class="icon-bg-light icon-twitter icon-circled icon-1x"></i></a>
                        </li>
                        <li><a href="www.linkedin.com/in/julianacostasilva" title="Linkedin"><i
                                class="icon-bg-light icon-linkedin icon-circled icon-1x"></i></a></li>
                        <li><a href="https://github.com/costasilvati/consexpression" title="Linkedin"><i
                                class="icon-bg-light icon-github icon-circled icon-1x"></i></a></li>

                    </ul>
                </div>
            </div>
        </div>
    </div>
    <div class="subfooter">
        <div class="container">
            <div class="row">
                <div class="span6">
                    <p>
                        &copy; Plato - All right reserved
                    </p>
                </div>
                <div class="span6">
                    <div class="pull-right">
                        <div class="credits">
                            <!--
                              All the links in the footer should remain intact.
                              You can delete the links only if you purchased the pro version.
                              Licensing information: https://bootstrapmade.com/license/
                              Purchase the pro version with working PHP/AJAX contact form: https://bootstrapmade.com/buy/?theme=Plato
                            -->
                            Designed by <a href="https://bootstrapmade.com/">BootstrapMade</a>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    </div>
</footer>

<script src="assets/js/jquery.js"></script>
<script src="assets/js/modernizr.js"></script>
<script src="assets/js/jquery.easing.1.3.js"></script>
<script src="assets/js/google-code-prettify/prettify.js"></script>
<script src="assets/js/bootstrap.js"></script>
<script src="assets/js/jquery.prettyPhoto.js"></script>
<script src="assets/js/portfolio/jquery.quicksand.js"></script>
<script src="assets/js/portfolio/setting.js"></script>
<script src="assets/js/hover/jquery-hover-effect.js"></script>
<script src="assets/js/jquery.flexslider.js"></script>
<script src="assets/js/classie.js"></script>
<script src="assets/js/cbpAnimatedHeader.min.js"></script>
<script src="assets/js/jquery.refineslide.js"></script>
<script src="assets/js/jquery.ui.totop.js"></script>

<!-- Template Custom Javascript File -->
<script src="assets/js/custom.js"></script>

</body>

</html>
